### PR TITLE
Adding hardware test

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Adding hardware test

## Motivation and Context

- Adding hardware test
- Run real hardware tests on real devices!

## How Has This Been Tested?

With this PR pipeline

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

## Summary by CodeRabbit

- **New Features**
	- Enhanced Azure DevOps pipeline configuration with new triggers and resource management.
	- Introduced a `Build` stage and job for improved build processes.
	- Added dynamic testing capabilities with the `appComponents` parameter.

- **Bug Fixes**
	- Implemented auto-cancellation for previous builds on new commits in pull requests.

- **Documentation**
	- Updated pipeline configurations to exclude specific paths from triggering builds.

### CODE REVIEW

**Comment by Ellerbach** [comment]
Couple of things to improve in the device-test.yml template:
* catch more issues like "No test is available in " see https://dev.azure.com/nanoframework/nanoFramework.Json/_build/results?buildId=49956&view=logs&j=e58a6411-6966-50fc-20aa-ceb63c9d7583&t=0a4b724a-a371-5bca-75e7-da3b7d83eb80
* Adjust the adapter for the path when on Linux: https://dev.azure.com/nanoframework/nanoFramework.Json/_build/results?buildId=49957&view=logs&j=e58a6411-6966-50fc-20aa-ceb63c9d7583&t=0a4b724a-a371-5bca-75e7-da3b7d83eb80

File: azure-bootstrap.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  autoCancel: true
+
+# add nf-tools repo to resources (for Azure Pipelines templates)
+resources:
+  repositories:
+    - repository: templates
+      type: github
+      name: nanoframework/nf-tools
+      endpoint: nanoframework
+
```
[the remark this answers is not in the document]
Comment (Ellerbach) [reply azure-bootstrap.yml:34]: this is how we're using it

**Comment by josesimoes** [comment]
@nfbot updatedependencies

**Review by josesimoes — APPROVED** [review]
LGTM! Another one bites the dust...

File: azure-bootstrap.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  displayName: Trigger Azure Dev Ops build and test pipeline
+  pool:
+    vmImage: 'ubuntu-latest'  
+
+
+  steps:
+  - template: azure-pipelines-templates/device-bootstrap.yml@templates
+    parameters:
+      AZURE_DEVOPS_PROJECT: nanoFramework.Json
+      AZURE_DEVOPS_PIPELINE_ID: 59
```
[the remark this answers is not in the document]
Comment (josesimoes) [reply azure-bootstrap.yml]: retry handled in the test runner

**Review by josesimoes — COMMENTED** [review]
Can't get much better than this. Done testing improvements in the playground repo.
All good to merge!

**Comment by josesimoes** [comment]
@nfbot updatedependencies

**Comment by Ellerbach** [comment]
Merging as beeing tested and nf-tools being adjusted

**Pull request merged**